### PR TITLE
Fix invalid WKT empty geometries

### PR DESCRIPTION
**Pull request by Oreilles**
Currently, empty geometries would be writter in WKT as:
```
LINESTRING()
POLYGON()
GEOMETRYCOLLECTION()
```
which isn't valid WKT.

This PR addresses this issue and output instead:
```
LINESTRING EMPTY
POLYGON EMPTY
GEOMETRYCOLLECTION EMPTY
```

### CODE REVIEW

File: geozero/src/wkt/wkt_writer.rs
Comment (michaelkirk) [inline geozero/src/wkt/wkt_writer.rs:75]: Would we ever expect this to be `None`?

**Review by michaelkirk — APPROVED** [review]
Good catch!

File: geozero/src/wkt/wkt_writer.rs
Comment (michaelkirk) [inline geozero/src/wkt/wkt_writer.rs:123]: This seems less weird. 👍

File: geozero/src/wkt/wkt_writer.rs
Comment (Oreilles) [reply geozero/src/wkt/wkt_writer.rs:75]: A geometry would have to end before it began, so in theory no ?

Comment (michaelkirk) [reply geozero/src/wkt/wkt_writer.rs:75]: Consider adding something like `else { debug_assert!(false, "ended geometry that didn't start") }` as a "this should never happen but we don't want to blow up in production if it does"

Comment (Oreilles) [reply geozero/src/wkt/wkt_writer.rs:75]: Done 👍

**Comment by Oreilles** [comment]
We have a problem with the csv `BufferingWktWriter`... since it instanciate a new writer for every processor function call, the state isn't persisted and the implementation cannot work.

File: geozero/src/wkt/wkt_writer.rs
Comment (Oreilles) [reply geozero/src/wkt/wkt_writer.rs:75]: Nice catch... https://github.com/georust/geozero/pull/160#issuecomment-1676431784

**Comment by michaelkirk** [comment]
> We have a problem with the csv BufferingWktWriter... since it instanciate a new writer for every processor function call, the state isn't persisted and the implementation cannot work.

I've just opened https://github.com/Oreilles/geozero/pull/3 against your own branch that I think addresses this (and simplifies my old crappy code from yesteryear!). Take a look and merge into your branch if you think it makes sense.

**Comment by michaelkirk** [comment]
> If we had access to the geometry size in processor functions polygon_end etc., then we wouldn't have to manage the geometry_sizes stack 🤔

I'm also open to this approach (or we could do both) but it seems like a bigger breaking change, so my preference at this point would be to avoid it.

File: geozero/src/wkt/wkt_writer.rs
Comment (michaelkirk) [inline geozero/src/wkt/wkt_writer.rs:20]: This LGTM, but wanted to highlight the change for other reviewers that we're changing WktWriter to take any `impl Write` rather than only `&mut impl Write`.

Comment (nyurik) [reply geozero/src/wkt/wkt_writer.rs:20]: I'm all for simplifying writters so that they own the underlying `impl Write` - but should it possibly be a separate PR that changes them all?  It feels like a counter-pattern to introduce it in just one writer as part of an unrelated empty geometry support. Then we can rebase this PR on top of it?

Comment (pka) [reply geozero/src/wkt/wkt_writer.rs:20]: I'm fine with doing it for all writers. Who wants?

Comment (nyurik) [reply geozero/src/wkt/wkt_writer.rs:20]: I'm ok either way, just as long as we don't release until all writers are more-or-less lined up (and of course its a minor rather than a patch change

Comment (michaelkirk) [reply geozero/src/wkt/wkt_writer.rs:20]: > and of course its a minor rather than a patch change

Just to be clear, changing from `&mut Write -> Write` is a non-breaking change, right?

Comment (michaelkirk) [reply geozero/src/wkt/wkt_writer.rs:20]: Yes, there is a breaking change to WktWriter in this PR if that's what you mean (see the changelog entry).

So is everyone OK to merge this?

**Review by nyurik — APPROVED** [review]
lgtm

**Comment by nyurik** [comment]
i thought there were some discussions in rust itself to stop using bors because github now covers most such usecases?

**Comment by michaelkirk** [comment]
Yes, there were some of us trying to figure out GH merge queues a couple weeks ago. If you look at the proj crate’s recent commits for example, you’ll see some of my misadventures. As far as I know we don’t have any working examples on georust where GH’s merge queues function like bors: i.e. I trigger the process, then the tests run against an up-to-date branch and only get merged into main once they pass.It seems so basic, but i kept finding that the branches would merge immediately, regardless of if the tests pass.I’m sure it’s possible, I just didn’t personally dedicate the time to figuring it out. If you tell me what series of buttons to press I can press them. On Aug 16, 2023, at 12:57, Yuri Astrakhan ***@***.***> wrote:﻿
i thought there were some discussions in rust itself to stop using bors because github now covers most such usecases?

**Comment by michaelkirk** [comment]
I just realized that this project does not, and never did, use bors. 😅

**Comment by nyurik** [comment]
@michaelkirk in Martin, we have branch protection rule for `main`:
* Require a pull request before merging
* Require approvals (1)
* Require status checks to pass before merging (with most of the key CI jobs listed as required)

I think if you also enable `Require branches to be up to date before merging`, you will achieve what you want:  a way to +1 something, plus a way to say "squash once CI passes" - this way it will not be squashed unless everything is good, and it will not pass if another PR merges ahead of it.  Granted that it will fail if another PR merges first though.

**Comment by michaelkirk** [comment]
I got it working! Ultimately I think the issue was [something else](https://github.com/georust/proj/pull/179#issuecomment-1684521823), but thanks for the sanity check @nyurik.

**Pull request merged**
